cmd/scip/test.go

```diff
@@ -0,0 +1,369 @@
+			}
+			result = append(result, &symbolAttribute{
+				start:          start,
+				length:         length,
+				kind:           kind,
+				data:           occ.Symbol,
+				additionalData: []string{},
+			})
+
+			for _, diagnostic := range occ.Diagnostics {
+				result = append(result, &symbolAttribute{
+					start:  start,
+					length: length,
+					kind:   "diagnostic",
+					data:   diagnostic.Severity.String(),
+					additionalData: []string{
+						diagnostic.Message,
+					},
+				})
+			}
+		}
+	}
+	return result
+}
+
+func parseTestCase(line string, leadingLines []string, commentSyntax string) *symbolAttributeTestCase {
+	start := 0
+	length := 0
+	enforceLength := false
+
+	if strings.Contains(line, "<-") {
+		// if the test line selects via `<-`, treat the symbol selection
+		// as the location of the commentSyntax
+		start = strings.Index(line, commentSyntax)
+		line = strings.Replace(line, "<-", "", 1)
+	} else {
+		// otherwise treat the start as the first `^`
+		start = strings.Index(line, "^")
+
+		// a single `^` dictates no length enforcement
+		// anything more signifies length should be verified
+		if strings.Contains(line, "^^") {
+			enforceLength = true
+			length = charCountInString(line, '^')
+		}
+		line = strings.ReplaceAll(line, "^", "")
+	}
+
+	// remove the comment prefix & whitespace
+	line = strings.TrimSpace(strings.Replace(line, commentSyntax, "", 1))
+
+	// the type of the symbol should be the first word
+	// this is "definition", "reference", "documentation", "diagnostic", etc..
+	kind := strings.Split(line, " ")[0]
+
+	// the data is everything except the type
+	data := strings.TrimSpace(strings.Replace(line, kind, "", 1))
+
+	additionalData := []string{}
+	for i := range leadingLines {
+		leadingLine := leadingLines[i]
+		// if the leadingLine is not a comment line, we're outside of the test case block.
+		if !strings.HasPrefix(strings.TrimSpace(leadingLine), commentSyntax) {
+			break
+		}
+
+		// remove the comment character(s)
+		leadingLine = strings.Replace(leadingLine, commentSyntax, "", 1)
+
+		// if the leading line doesn't start with '>', its not a multiline case
+		if !strings.HasPrefix(strings.TrimSpace(leadingLine), ">") {
+			break
+		}
+
+		// remove the '>' character
+		leadingLine = strings.Replace(leadingLine, ">", "", 1)
+		additionalData = append(additionalData, strings.TrimSpace(leadingLine))
+	}
+
+	return &symbolAttributeTestCase{
+		attribute: &symbolAttribute{
+			kind:           kind,
+			start:          start,
+			length:         length,
+			data:           data,
+			additionalData: additionalData,
+		},
+		enforceLength: enforceLength,
+	}
+}
+
+func isValidTestCase(testCase *symbolAttributeTestCase, attributes []*symbolAttribute) bool {
+	for _, attr := range attributes {
+		if isValidTestCaseForAttribute(testCase, attr) {
+			return true
+		}
+	}
+	return false
+}
+
+func isValidTestCaseForAttribute(testCase *symbolAttributeTestCase, attr *symbolAttribute) bool {
```

Comment:
These should probably be methods on `symbolAttributeTestCase` called `CheckAll` and `Check` respectively.

The prefix 'isValid' makes it sound like we're doing a well-formedness check, whereas we're actually doing a _comparison_ between the SCIP index and the test data.